### DISCUSSION TITLE
Can CI runs and known bots be excluded from increasing counter on crates.io?

### DISCUSSION THREAD

**Discussion by toni-rmc** [rust-lang/crates.io#8913] [General]
I have crate published and would like to know number of real users to determine how much time to spend working on the crate and basically to see if crate is actually used.
I see pretty much all versions, old and new downloaded almost the same number of times, except when I publish a new version.

As I understand it CI runs are counted, as well as some known bots. Can those be identified or required to identify themselves as bots and CI runs so they are not counted?

Yes I know single user can download one crate more than once so number of downloads is not equal people using it, and that is fine, it still would give much better statistic. Single user can use same crate in multiple projects.

Also most bot downloads come from well known bots, like docs.rs, lib.rs.

So can CI runs and known bots be ignored because why have counter at all if it's pretty much useless for statistics?
What is the point of having a counter on crates.io if it doesn't provide useful information to crate authors?

**Comment by Turbo87** [comment]
the problem is that there is no reliable way of detecting CI systems and other bots. we discussed this in the past and decided that it would be better to count all downloads instead of trying to detect CI systems but then only catching a subset of them, which skews the numbers in an even worse way.

**Reply by toni-rmc** [reply]
But as I understood, the biggest culprits by far are CI runs of Rust itself and known bots like docs.rs and lib.rs, can't they sent some kind of header to identify them as bots. Remove them and count number is way more accurate. Sure there are others out there but I think their numbers are negligible. Its not like many project out there that do regular runs of the entire crates.io.

Those make by far the largest subset of them, if I would have to guess over 98%, and catching them would give way more accurate representation of real crate usage how that can be the worse?
Counting them, when can they easily be excluded literary makes counter useless.
